README.md
# godot-snake

Godot snake


CHANGELOG.md
# Change Log

All notable changes to this project will be documented in this file. See [standard-version](https://github.com/conventional-changelog/standard-version) for commit guidelines.

<a name="0.4.0"></a>
# [0.4.0](https://github.com/ilyaev/godot-snake/compare/v0.3.0...v0.4.0) (2018-04-21)


### Bug Fixes

* default DPAD ([97a040a](https://github.com/ilyaev/godot-snake/commit/97a040a))


### Features

* Release One ([353f794](https://github.com/ilyaev/godot-snake/commit/353f794))



<a name="0.3.0"></a>
# [0.3.0](https://github.com/ilyaev/godot-snake/compare/v0.2.3...v0.3.0) (2018-04-16)


### Features

* updated AI models ([1cd0167](https://github.com/ilyaev/godot-snake/commit/1cd0167))



<a name="0.2.3"></a>
## [0.2.3](https://github.com/ilyaev/godot-snake/compare/v0.2.2...v0.2.3) (2018-04-16)



<a name="0.2.2"></a>
## [0.2.2](https://github.com/ilyaev/godot-snake/compare/v0.2.1...v0.2.2) (2018-04-16)



<a name="0.2.1"></a>
## [0.2.1](https://github.com/ilyaev/godot-snake/compare/v0.2.0...v0.2.1) (2018-04-16)



<a name="0.2.0"></a>
# [0.2.0](https://github.com/ilyaev/godot-snake/compare/v0.1.19...v0.2.0) (2018-04-16)


### Features

* release one ([909c722](https://github.com/ilyaev/godot-snake/commit/909c722))



<a name="0.1.19"></a>
## [0.1.19](https://github.com/ilyaev/godot-snake/compare/v0.1.18...v0.1.19) (2018-04-16)



<a name="0.1.18"></a>
## [0.1.18](https://github.com/ilyaev/godot-snake/compare/v0.1.17...v0.1.18) (2018-04-14)


### Bug Fixes

* http thread improvements ([c98949d](https://github.com/ilyaev/godot-snake/commit/c98949d))



<a name="0.1.17"></a>
## [0.1.17](https://github.com/ilyaev/godot-snake/compare/v0.1.0...v0.1.17) (2018-04-14)


### Bug Fixes

* better multithreading for ai. build management ([d0c1f6a](https://github.com/ilyaev/godot-snake/commit/d0c1f6a))



<a name="0.1.0"></a>
# 0.1.0 (2018-04-13)


### Bug Fixes

* resource optimization. hud improvements ([d7e8402](https://github.com/ilyaev/godot-snake/commit/d7e8402))
* **ai:** pathfinding fix. introduce tweening instead of custom stuff. ([3b98103](https://github.com/ilyaev/godot-snake/commit/3b98103))
* **ai:** Rerouting. Closes [#5](https://github.com/ilyaev/godot-snake/issues/5) ([2f4179d](https://github.com/ilyaev/godot-snake/commit/2f4179d))
* control spaming no more ([931c4cb](https://github.com/ilyaev/godot-snake/commit/931c4cb))
* invincible mode dont ruin walls. Closes  [#18](https://github.com/ilyaev/godot-snake/issues/18) ([c8e848b](https://github.com/ilyaev/godot-snake/commit/c8e848b))
* Memory leaks optimized ([771d89e](https://github.com/ilyaev/godot-snake/commit/771d89e))
* memory management. better explosions. snake destroy sequence. ([2021adc](https://github.com/ilyaev/godot-snake/commit/2021adc))
* movements are now smooth ([e2531c7](https://github.com/ilyaev/godot-snake/commit/e2531c7))
* smooth motion update ([ab8b7eb](https://github.com/ilyaev/godot-snake/commit/ab8b7eb))
* snake tail, explosions synchronize. optimization closes [#4](https://github.com/ilyaev/godot-snake/issues/4) ([eaa2e45](https://github.com/ilyaev/godot-snake/commit/eaa2e45))
* sync changes. better vortex sequence ([9f22fd4](https://github.com/ilyaev/godot-snake/commit/9f22fd4))
* tail wall fix ([6f7a770](https://github.com/ilyaev/godot-snake/commit/6f7a770))


### Features

* academy ([caeba75](https://github.com/ilyaev/godot-snake/commit/caeba75))
* academy builder. DQN agent improvements. ([6303687](https://github.com/ilyaev/godot-snake/commit/6303687))
* autostart snake after 3 seconds. closes [#28](https://github.com/ilyaev/godot-snake/issues/28) ([c23a6a8](https://github.com/ilyaev/godot-snake/commit/c23a6a8))
* debug popup menu ([707a6ef](https://github.com/ilyaev/godot-snake/commit/707a6ef))
* different color schema for enemy stuff. Closes [#7](https://github.com/ilyaev/godot-snake/issues/7) ([6a5b963](https://github.com/ilyaev/godot-snake/commit/6a5b963))
* dpad ([d6a8cd0](https://github.com/ilyaev/godot-snake/commit/d6a8cd0))
* DQN AI agent works now. Closes [#20](https://github.com/ilyaev/godot-snake/issues/20) ([f323dd2](https://github.com/ilyaev/godot-snake/commit/f323dd2))
* fade and flash effects ([c238a7c](https://github.com/ilyaev/godot-snake/commit/c238a7c))
* final levels design. closes [#30](https://github.com/ilyaev/godot-snake/issues/30) ([e1aaa80](https://github.com/ilyaev/godot-snake/commit/e1aaa80))
* food animation improvement. closes [#27](https://github.com/ilyaev/godot-snake/issues/27) ([7a9f7c5](https://github.com/ilyaev/godot-snake/commit/7a9f7c5))
* game over scene transition ([536a0b6](https://github.com/ilyaev/godot-snake/commit/536a0b6))
* game over screen have score ([acf95d0](https://github.com/ilyaev/godot-snake/commit/acf95d0))
* Game settings panel. Start scene layout fixes ([d218627](https://github.com/ilyaev/godot-snake/commit/d218627))
* gameover scene improvement ([fb0e1a9](https://github.com/ilyaev/godot-snake/commit/fb0e1a9))
* gameover sequene placeholder and state ([869854a](https://github.com/ilyaev/godot-snake/commit/869854a))
* global scene manager, game state patter, snake death/birth animation control ([aeabba1](https://github.com/ilyaev/godot-snake/commit/aeabba1))
* grpahics updated ([490fadd](https://github.com/ilyaev/godot-snake/commit/490fadd))
* Hall of fame. in progress ([52fd46d](https://github.com/ilyaev/godot-snake/commit/52fd46d))
* introduce fruit action. fixed enemy snake spawner. ([e2d7480](https://github.com/ilyaev/godot-snake/commit/e2d7480))
* Introduce HUD. Score widget. Closes [#9](https://github.com/ilyaev/godot-snake/issues/9) ([f227146](https://github.com/ilyaev/godot-snake/commit/f227146))
* introduce new snake state - Flash. Closes [#19](https://github.com/ilyaev/godot-snake/issues/19) ([70e6b2b](https://github.com/ilyaev/godot-snake/commit/70e6b2b))

export/levels.js
module.exports = {
    levels: [
        { walls: [], maxFood: 1, maxX: 8, maxY: 8, pits: [], name: 'empty8x8' },
        { walls: [], maxFood: 4, maxX: 15, maxY: 15, pits: [], name: 'empty16x16' },
        { walls: [], maxFood: 8, maxX: 32, maxY: 32, pits: [], name: 'empty32x32' },
        {
            walls: [
                { x: 0, y: 0 },
                { x: 0, y: 1 },
                { x: 0, y: 2 },
                { x: 0, y: 18 },
                { x: 0, y: 19 },
                { x: 0, y: 20 },
                { x: 1, y: 0 },
                { x: 1, y: 1 },
                { x: 1, y: 19 },
                { x: 1, y: 20 },
                { x: 2, y: 0 },
                { x: 2, y: 20 },
                { x: 5, y: 5 },
                { x: 5, y: 6 },
                { x: 5, y: 7 },
                { x: 5, y: 8 },
                { x: 5, y: 9 },
                { x: 5, y: 10 },
                { x: 5, y: 11 },
                { x: 5, y: 12 },
                { x: 5, y: 13 },
                { x: 5, y: 14 },
                { x: 5, y: 15 },
                { x: 11, y: 8 },
                { x: 11, y: 12 },
                { x: 12, y: 8 },
                { x: 12, y: 12 },
                { x: 13, y: 8 },
                { x: 13, y: 12 },
                { x: 14, y: 8 },
                { x: 14, y: 12 },
                { x: 15, y: 8 },
                { x: 15, y: 12 },
                { x: 16, y: 8 },
                { x: 16, y: 12 },
                { x: 22, y: 5 },
                { x: 22, y: 6 },
                { x: 22, y: 7 },
                { x: 22, y: 8 },
                { x: 22, y: 9 },
                { x: 22, y: 10 },
                { x: 22, y: 11 },
                { x: 22, y: 12 },
                { x: 22, y: 13 },
                { x: 22, y: 14 },
                { x: 22, y: 15 },
                { x: 24, y: 0 },
                { x: 24, y: 20 },
                { x: 25, y: 0 },
                { x: 25, y: 1 },
                { x: 25, y: 19 },
                { x: 25, y: 20 },
                { x: 26, y: 0 },
                { x: 26, y: 1 },
                { x: 26, y: 2 },
                { x: 26, y: 18 },
                { x: 26, y: 19 },
                { x: 26, y: 20 }
            ],
            maxFood: 5,
            maxX: 27,
            maxY: 21,
            pits: [],
            name: 'one'
        },
        {
            walls: [
                { x: 2, y: 6 },
                { x: 3, y: 6 },
                { x: 4, y: 2 },
                { x: 4, y: 3 },
                { x: 4, y: 4 },
                { x: 4, y: 6 },
                { x: 5, y: 6 },
                { x: 6, y: 6 },
                { x: 7, y: 6 },
                { x: 8, y: 6 },
                { x: 9, y: 6 },
                { x: 10, y: 6 },
                { x: 10, y: 8 },
                { x: 10, y: 9 },
                { x: 10, y: 10 },
                { x: 11, y: 6 },
                { x: 12, y: 6 }
            ],
            maxFood: 6,
            maxX: 15,
            maxY: 15,
            pits: [],
            name: 'two'
        },
        {
            walls: [
                { x: 3, y: 8 },
                { x: 3, y: 9 },
                { x: 3, y: 10 },
                { x: 3, y: 11 },
                { x: 3, y: 12 },
                { x: 3, y: 13 },
                { x: 4, y: 13 },
                { x: 5, y: 13 },
                { x: 6, y: 6 },
                { x: 6, y: 13 },
                { x: 7, y: 6 },
                { x: 7, y: 13 },
                { x: 8, y: 6 },
                { x: 9, y: 6 },
                { x: 10, y: 6 },
                { x: 10, y: 7 },
                { x: 10, y: 8 },
                { x: 10, y: 9 },
                { x: 10, y: 10 }
            ],
            maxFood: 7,
            maxX: 15,
            maxY: 15,
            pits: [],
            name: 'three'
        },
        {
            walls: [
                { x: 5, y: 16 },
                { x: 5, y: 17 },
                { x: 5, y: 18 },
                { x: 5, y: 19 },
                { x: 5, y: 20 },
                { x: 6, y: 6 },
                { x: 6, y: 7 },
                { x: 6, y: 8 },
                { x: 6, y: 9 },
                { x: 6, y: 10 },
                { x: 6, y: 29 },
                { x: 7, y: 29 },
                { x: 8, y: 29 },
                { x: 9, y: 2 },
                { x: 9, y: 16 },
                { x: 9, y: 29 },
                { x: 10, y: 2 },
                { x: 10, y: 16 },
                { x: 10, y: 29 },
                { x: 11, y: 2 },
                { x: 11, y: 16 },
                { x: 12, y: 2 },
                { x: 12, y: 16 },
                { x: 12, y: 24 },
                { x: 13, y: 2 },
                { x: 13, y: 16 },
                { x: 13, y: 24 },
                { x: 14, y: 24 },
                { x: 15, y: 24 },
                { x: 16, y: 8 },
                { x: 16, y: 9 },
                { x: 16, y: 10 },
                { x: 16, y: 11 },
                { x: 16, y: 12 },
                { x: 16, y: 13 },
                { x: 16, y: 14 },
                { x: 16, y: 15 },
                { x: 16, y: 24 },
                { x: 17, y: 24 },
                { x: 18, y: 24 },
                { x: 22, y: 5 },
                { x: 23, y: 5 },
                { x: 24, y: 5 },
                { x: 24, y: 25 },
                { x: 24, y: 26 },
                { x: 24, y: 27 },
                { x: 24, y: 28 },
                { x: 25, y: 5 },
                { x: 26, y: 5 },
                { x: 26, y: 16 },
                { x: 26, y: 17 },
                { x: 26, y: 18 },
                { x: 26, y: 19 },
                { x: 26, y: 20 },
                { x: 26, y: 21 },
                { x: 27, y: 5 }
            ],
            maxFood: 8,
            maxX: 32,
            maxY: 32,
            pits: [],
            name: 'four'
        },
        {
            walls: [
                { x: 2, y: 3 },
                { x: 3, y: 3 },
                { x: 4, y: 3 },
                { x: 5, y: 2 },
                { x: 5, y: 3 },
                { x: 5, y: 16 },
                { x: 5, y: 17 },
                { x: 5, y: 18 },
                { x: 5, y: 19 },
                { x: 5, y: 23 },
                { x: 6, y: 2 },
                { x: 6, y: 19 },
                { x: 6, y: 23 },
                { x: 6, y: 27 },
                { x: 7, y: 2 },
                { x: 7, y: 3 },
                { x: 7, y: 4 },
                { x: 7, y: 5 },
                { x: 7, y: 6 },
                { x: 7, y: 12 },
                { x: 7, y: 19 },
                { x: 7, y: 23 },
                { x: 7, y: 27 },
                { x: 8, y: 12 },
                { x: 8, y: 13 },
                { x: 8, y: 14 },
                { x: 8, y: 15 },
                { x: 8, y: 16 },
                { x: 8, y: 19 },
                { x: 8, y: 23 },
                { x: 8, y: 27 },
                { x: 9, y: 12 },
                { x: 9, y: 13 },
                { x: 9, y: 14 },
                { x: 9, y: 15 },
                { x: 9, y: 16 },
                { x: 9, y: 17 },
                { x: 9, y: 18 },
                { x: 9, y: 19 },
                { x: 9, y: 23 },
                { x: 9, y: 27 },
                { x: 10, y: 27 },
                { x: 11, y: 27 },
                { x: 12, y: 22 },
                { x: 12, y: 23 },
                { x: 12, y: 24 },
                { x: 12, y: 25 },
                { x: 12, y: 26 },
                { x: 12, y: 27 },
                { x: 12, y: 28 },
                { x: 12, y: 29 },
                { x: 12, y: 30 },
                { x: 14, y: 16 },
                { x: 15, y: 12 },
                { x: 16, y: 10 },
                { x: 16, y: 11 },
                { x: 16, y: 12 },
                { x: 17, y: 10 },
                { x: 17, y: 11 },
                { x: 17, y: 12 },
                { x: 18, y: 10 },
                { x: 18, y: 11 },
                { x: 18, y: 12 },
                { x: 19, y: 10 },
                { x: 19, y: 19 },
                { x: 19, y: 20 },
                { x: 19, y: 21 },
                { x: 19, y: 22 },
                { x: 19, y: 23 },
                { x: 19, y: 24 },
                { x: 20, y: 19 },
                { x: 20, y: 20 },
                { x: 20, y: 21 },
                { x: 20, y: 22 },
                { x: 20, y: 23 },
                { x: 20, y: 24 },
                { x: 23, y: 12 },
                { x: 24, y: 12 },
                { x: 25, y: 5 },
                { x: 25, y: 6 },
                { x: 25, y: 7 },
                { x: 25, y: 12 },
                { x: 26, y: 5 },
                { x: 26, y: 12 },
                { x: 26, y: 22 },
                { x: 26, y: 23 },
                { x: 26, y: 24 },
                { x: 26, y: 25 },
                { x: 27, y: 5 },
                { x: 27, y: 12 },
                { x: 28, y: 5 },
                { x: 28, y: 6 },
                { x: 28, y: 7 },
                { x: 28, y: 8 },
                { x: 28, y: 9 },
                { x: 28, y: 10 },
                { x: 28, y: 11 },
                { x: 28, y: 12 },
                { x: 28, y: 13 },
                { x: 28, y: 14 },
                { x: 28, y: 15 }
            ],
            maxFood: 9,
            maxX: 32,
            maxY: 32,
            pits: [],
            name: 'five'
        }
    ]
}
